scripts/asm-installer/install_asm

```diff
@@ -707,27 +713,121 @@ check_istio_not_deployed(){
 
 validate_istio_version() {
   info "Checking existing Istio version(s)..."
-  local OUTPUT
-  OUTPUT="$(retry 3 ./istio-${RELEASE}/bin/istioctl version -o json)"
-  local FOUND_VALID_VERSION
-  FOUND_VALID_VERSION="false"
-  for version in $(echo "${OUTPUT}" | jq -r '.meshVersion[].Info.version' -r); do
+  local VERSION_OUTPUT; VERSION_OUTPUT="$(retry 3 ./${ISTIOCTL_REL_PATH} version -o json)"
+  if [[ -z "${VERSION_OUTPUT}" ]]; then
+    fatal "Couldn't validate existing Istio versions."
+  fi
+  local FOUND_VALID_VERSION; FOUND_VALID_VERSION=0
+  for version in $(echo "${VERSION_OUTPUT}" | jq -r '.meshVersion[].Info.version' -r); do
     if [[ "$version" =~ ^$RELEASE_LINE ]]; then
       info "  $version (suitable for migration)"
-      FOUND_VALID_VERSION="true"
-    fi
-    if ! [[ "$version" =~ ^$RELEASE_LINE ]]; then
+      FOUND_VALID_VERSION=1
+    else
       info "  $version (not suitable for migration)"
     fi
     if [[ "$version" =~ "asm" ]]; then
       fatal "Cannot migrate from version $version. Only migration from OSS Istio to the ASM distribution is supported."
     fi
   done
-  if [[ "$FOUND_VALID_VERSION" != "true" ]]; then
+  if [[ "$FOUND_VALID_VERSION" -eq 0 ]]; then
     fatal "Migration requires an existing control plane in the ${RELEASE_LINE} line."
   fi
 }
 
+validate_asm_version() {
+  info "Checking existing ASM version(s)..."
+  local VERSION_OUTPUT; VERSION_OUTPUT="$(retry 3 ./${ISTIOCTL_REL_PATH} version -o json)"
+  if [[ -z "${VERSION_OUTPUT}" ]]; then
+    fatal "Couldn't validate existing Istio versions."
+  fi
+  local FOUND_INVALID_VERSION; FOUND_INVALID_VERSION=0
+  for VERSION in $(echo "${VERSION_OUTPUT}" | jq -r '.meshVersion[].Info.version'); do
+    if ! [[ "${VERSION}" =~ "asm" ]]; then
+      fatal "Cannot upgrade from version ${VERSION}. Only upgrades from ASM distributions are supported."
+    fi
+
+    if version_valid_for_upgrade "${VERSION}"; then
+      info "  ${VERSION} (suitable for migration)"
+    else
+      info "  ${VERSION} (not suitable for migration)"
+      FOUND_INVALID_VERSION=1
+    fi
+  done
+  if [[ "$FOUND_INVALID_VERSION" -eq 1 ]]; then
+    fatal "Upgrade requires all existing control planes to be between versions 1.$((MINOR-1)).0 (inclusive) and ${RELEASE} (exclusive)."
+  fi
+}
+
+version_valid_for_upgrade() {
+  local VERSION; VERSION=$1
+  
+  # if asm version found, pattern: 1.6.11-asm.1-586f900508ad482ed32b830dd15f6c54b32b93ed
+  local VERSION_MAJOR VERSION_MINOR VERSION_POINT VERSION_REV
+  IFS="." read -r VERSION_MAJOR VERSION_MINOR VERSION_POINT VERSION_REV <<EOF
+${VERSION}
+EOF
+  VERSION_POINT="$(sed 's/-.*//' <<EOF
+${VERSION_POINT}
+EOF
+)"
+  VERSION_REV="$(sed 's/-.*//' <<EOF
+${VERSION_REV}
+EOF
+)"
+  if is_major_minor_invalid || is_minor_point_rev_invalid; then
+    return 1
+  else
+    return 0
+  fi
+}
+
+is_major_minor_invalid() {
+  [[ "$VERSION_MAJOR" -ne 1 ]] && return 0
```

Comment:
It looks like the VERSION_* variables were marked local, does this work?

Reply:
It works with bash due to dynamic scoping.